### PR TITLE
Balance open polls across TQ partition

### PR DESCRIPTION
**Pull request by ShahabT**
**What changed?**
For each TQ, Matching client LB keeps track of open pollers per partition and selects a partition with the fewest open pollers. Forwarded polls are only counted once, for the original partition.

**Why?**
In situation where a TQ has no task to dispatch (or it cannot dispatch for another reason) it can become a black hole for pollers. This change prevent that to happen (assuming the number of pollers is greater than the number of partitions)

**How did you test it?**
Unit test
Locally ran against multiple workloads

**Potential risks**
None known

**Is hotfix candidate?**
No

### CODE REVIEW

File: client/matching/loadbalancer.go [hunk 1]
```diff
@@ -135,6 +140,130 @@ func (lb *defaultLoadBalancer) pickPartition(
...
+
+	return &pollToken{
+		fullName:    b.taskQueue.Name.WithPartition(partitionID).FullName(),
+		partitionID: partitionID,
+		balancer:    b,
+	}
+}
+
+// caller to ensure that lock is obtained before call this function
+func (b *tqLoadBalancer) pickReadPartitionWithFewestPolls(partitionCount int) int {
```
Comment (tdeebswihart) [inline client/matching/loadbalancer.go:218]: How many partitions can we have? If the answer is "a lot" we may want to look at a randomized approach; something like https://www.nginx.com/blog/nginx-power-of-two-choices-load-balancing-algorithm/ ([original paper](https://www.eecs.harvard.edu/~michaelm/postscripts/tpds2001.pdf))

Comment (ShahabT) [reply client/matching/loadbalancer.go:218]: Thanks @tdeebswihart, I learned this today. I feel what we have is ok, at least for now. almost all use cases should be fine with a few partitions. I don't imagine anyone needs more than a 100.

Comment (tdeebswihart) [reply client/matching/loadbalancer.go:218]: Do we have data from a production cluster for how many task queue partitions we have? Say, just the max partitions across all task queues

Comment (tdeebswihart) [reply client/matching/loadbalancer.go:218]: Yimin replied with `256` in Slack

**Comment by tdeebswihart** [comment]
> In situation where a TQ has no task to dispatch (or it cannot dispatch for another reason) it can become a black hole for pollers.

How does that happen, and how does this fix it?

**Comment by ShahabT** [comment]
> > In situation where a TQ has no task to dispatch (or it cannot dispatch for another reason) it can become a black hole for pollers.
> 
> How does that happen, and how does this fix it?

Say there are N partitions, and only one has tasks. A poller either long polls on a wrong partition (stuck for a minute), or goes to the right one and immediately receives a task, and comes for another one and with (N-1)/N probablity goes to a wrong partition. repeated enough times, all pollers find themselves stuck on one of the empty partitions for a minute. the task dispatch throughput will be pretty much limited to no. of pollers per minute.

This change prevents that because each time it sees that there is 1 poller hanging on each of the wrong partitions, so it'll send the rest of the pollers to the right partitions. so throughput will be (no. pollers minus no. partitions) per X ms, where X is the roundtrip of a single poller.

**Comment by tdeebswihart** [comment]
> > > In situation where a TQ has no task to dispatch (or it cannot dispatch for another reason) it can become a black hole for pollers.
> > 
> > 
> > How does that happen, and how does this fix it?
> 
> Say there are N partitions, and only one has tasks. A poller either long polls on a wrong partition (stuck for a minute), or goes to the right one and immediately receives a task, and comes for another one and with (N-1)/N probablity goes to a wrong partition. repeated enough times, all pollers find themselves stuck on one of the empty partitions for a minute. the task dispatch throughput will be pretty much limited to no. of pollers per minute.
> 
> This change prevents that because each time it sees that there is 1 poller hanging on each of the wrong partitions, so it'll send the rest of the pollers to the right partitions. so throughput will be (no. pollers minus no. partitions) per X ms, where X is the roundtrip of a single poller.

So this change better balances pollers across partitions but we could still have the case where the vast majority of pollers are on partitions with no work, yeah? We'd have an approximately even number of pollers per partition, but if only one partition has work then only `|pollers|/|partitions|` pollers will actually pick up work

File: client/matching/loadbalancer.go [hunk 2]
```diff
@@ -70,6 +71,28 @@ type (
 		nWritePartitions    dynamicconfig.IntPropertyFnWithTaskQueueInfoFilters
 		forceReadPartition  dynamicconfig.IntPropertyFn
 		forceWritePartition dynamicconfig.IntPropertyFn
+
+		lock         sync.Mutex
```
Comment (tdeebswihart) [inline client/matching/loadbalancer.go:75]: How often will we be writing to this map? Are we expecting a high write load? If not, we may wish to prefer an RWMutex (and we could just upgrade the lock if our entry is missing)

File: client/matching/loadbalancer.go [hunk 3]
```diff
@@ -135,6 +140,130 @@ func (lb *defaultLoadBalancer) pickPartition(
...
+	}
+}
+
+// caller to ensure that lock is obtained before call this function
+func (b *tqLoadBalancer) pickReadPartitionWithFewestPolls(partitionCount int) int {
+	// pick a random partition to start with
+	startPartitionID := rand.Intn(partitionCount)
+	pickedPartitionID := startPartitionID
+	minPollerCount := b.pollerCounts[pickedPartitionID]
+	for i := 1; i < partitionCount && minPollerCount > 0; i++ {
```
Comment (tdeebswihart) [inline client/matching/loadbalancer.go:223]: Is there any reason not to use a min-heap here? We also look to have a priority queue impl in our common package

**Review by tdeebswihart — COMMENTED** [review]
Shahab and I chatted about this offline. When a poller arrives and there's tasks waiting the poller is only on that partition for a few ms (vs a minute on an empty one). 

If, say, all pollers arrive at the same instant and only 1/N of partitions have tasks then most pollers will still become stuck on empty partitions, but it'll improve over time as the pollers on the task-filled partition won't stay long so it will likely have the fewest pollers at any point in time

File: client/matching/loadbalancer.go (hunk 2 above)
Comment (ShahabT) [reply client/matching/loadbalancer.go:75]: not much, whenever a TQ is seen for the first time in this process.

File: client/matching/loadbalancer.go (hunk 3 above)
Comment (ShahabT) [reply client/matching/loadbalancer.go:223]: The only reason would be simplicity I think. The min-heap implementation should have the following criteria to be a replacement:
- Ability to increase the priority of any given entry (needed for release)
- Ideally, randomized: when multiple minimums exist each should have the same probability of being popped 

Honestly, I'd not expect any min-heap implementation beat the runtime of a simple for loop for the small cardinality that we are looking at.

File: client/matching/loadbalancer.go (hunk 2 above)
Comment (ShahabT) [reply client/matching/loadbalancer.go]: good point. please review the new implementation.

File: client/matching/loadbalancer.go [hunk 4]
```diff
@@ -185,6 +177,28 @@ func (lb *defaultLoadBalancer) PickReadPartition(
 	return tqlb.pickReadPartition(partitionCount, lb.forceReadPartition())
 }
 
+func (lb *defaultLoadBalancer) getTaskQueLoadBalancer(
```
Comment (tdeebswihart) [inline client/matching/loadbalancer.go]:
```suggestion
func (lb *defaultLoadBalancer) getTaskQueueLoadBalancer(
```

File: client/matching/loadbalancer.go [hunk 5]
```diff
@@ -163,15 +163,7 @@ func (lb *defaultLoadBalancer) PickReadPartition(
...
-	key := taskQueueKey{NamespaceID: namespaceID, Name: parsedName, Type: taskQueueType}
-
-	lb.lock.Lock()
-	tqlb, ok := lb.taskQueueLBs[key]
-	if !ok {
-		tqlb = newTaskQueueLoadBalancer(key)
-		lb.taskQueueLBs[key] = tqlb
-	}
-	lb.lock.Unlock()
+	tqlb := lb.getTaskQueLoadBalancer(namespaceID, parsedName, taskQueueType)
```
Comment (tdeebswihart) [inline client/matching/loadbalancer.go]:
```suggestion
	tqlb := lb.getTaskQueueLoadBalancer(namespaceID, parsedName, taskQueueType)
```

**Review by tdeebswihart — APPROVED** [review]
Thank you!

File: client/matching/loadbalancer.go [hunk 6]
```diff
@@ -135,6 +140,144 @@ func (lb *defaultLoadBalancer) pickPartition(
...
+		lb.taskQueueLBs[key] = tqlb
+	}
+	lb.lock.Unlock()
+	return tqlb
+}
+
+func newTaskQueueLoadBalancer(key taskQueueKey) *tqLoadBalancer {
+	return &tqLoadBalancer{
+		taskQueue: key,
+		lock:      sync.Mutex{},
```
Comment (dnr) [inline client/matching/loadbalancer.go:205]: don't need this line, the zero value is fine

File: client/matching/loadbalancer.go [hunk 7]
```diff
@@ -135,6 +140,144 @@ func (lb *defaultLoadBalancer) pickPartition(
...
+			minPollerCount = b.pollerCounts[currPartitionID]
+		}
+	}
+
+	return pickedPartitionID
+}
+
+// caller to ensure that lock is obtained before call this function
+func (b *tqLoadBalancer) ensurePartitionCountLocked(partitionCount int) {
+	if len(b.pollerCounts) == int(partitionCount) {
```
Comment (dnr) [inline client/matching/loadbalancer.go:250]:
```suggestion
	if len(b.pollerCounts) == partitionCount {
```
and similar for the rest of this function

File: client/matching/loadbalancer.go [hunk 8]
```diff
@@ -135,6 +140,144 @@ func (lb *defaultLoadBalancer) pickPartition(
...
+	if err != nil || err == nil && !parsedName.IsRoot() {
+		// parse error or partition already picked, use as-is
+		return &pollToken{fullName: taskQueue.GetName()}
+	}
+
+	tqlb := lb.getTaskQueueLoadBalancer(namespaceID, parsedName, taskQueueType)
+
+	// For read path it's safer to return global default partition count instead of root partition, when we fail to
+	// map namespace ID to name.
+	var partitionCount = dynamicconfig.GlobalDefaultNumTaskQueuePartitions
```
Comment (dnr) [inline client/matching/loadbalancer.go:170]: :(  I don't really like any more code being aware of this default (that we hope to change eventually). What's not safe about returning 0? Tasks would get forwarded there eventually?

Comment (ShahabT) [reply client/matching/loadbalancer.go:170]: Tasks should forward at rate of 10 tasks per sec, but only if the root partition doesn't have a backlog.
This line should run very rarely (only when we fail to map NS name to ID). We can remove this later if needed with little/no trouble.

File: client/matching/loadbalancer.go (hunk 6 above)
Comment (ShahabT) [reply client/matching/loadbalancer.go:205]: addressed in #4981

File: client/matching/loadbalancer.go (hunk 7 above)
Comment (ShahabT) [reply client/matching/loadbalancer.go:250]: addressed in #4981

**Pull request merged**